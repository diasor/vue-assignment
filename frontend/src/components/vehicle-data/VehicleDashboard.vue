<template>
  <div class="main-container">
    <div class="vehicle-container">
      <div class="single-component">
        <google-map
          :key="mapChartKey"
          :currentPosition="currentVehicleData"
          :previousPositions="allMessages"
        />
      </div>
      <div class="single-component">
        <vehicle-data
          :speed="currentVehicleData.speed"
          :soc="currentVehicleData.soc"
          :energy="currentVehicleData.energy"
          :odo="currentVehicleData.odo"
        />
      </div>
    </div>
    <div class="vehicle-container">
      <div class="single-component">
        <line-chart
          :key="speedChartKey"
          title="Speed Profile"
          yTitle="Speed (km/h)"
          xTitle="Time"
          data-key="speed"
          :data="allMessages"
        />
      </div>
      <div class="single-component">
        <line-chart
          :key="socChartKey"
          title="State Of Charge Profile"
          yTitle="SoC (%)"
          xTitle="Time"
          data-key="soc"
          :data="allMessages"
        />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import { namespace } from "vuex-class";
import { Component } from "vue-property-decorator";
import GoogleMap from "@/components/maps/GoogleMap";
import VehicleData from "@/components/vehicle-data/VehicleData";
import LineChart from "@/components/charts/LineChart";

const SocketNameSpace = namespace("webSocketState/");
@Component({
  name: "VehicleDashboard",
  components: {
    googleMap: GoogleMap,
    vehicleData: VehicleData,
    lineChart: LineChart
  }
})
export default class VehicleDashboard extends Vue {
  @SocketNameSpace.Getter("currentVehicleData") currentVehicleData!: any;
  @SocketNameSpace.Getter("messages") messages!: any;

  timeout = 0;
  lastKey: number = 0;
  speedChartKey: string = "speedChartKey";
  socChartKey: string = "socChartKey";
  mapChartKey: string = "mapChartKey";
  windowSize: number = 0;

  get vehicleData() {
    return this.currentVehicleData;
  }

  get allMessages() {
    return this.messages;
  }

  reloadComponents() {
    this.lastKey = this.lastKey + 1;
    this.mapChartKey = `mapChartKey_${this.lastKey}`;
    this.speedChartKey = `speedChartKey_${this.lastKey}`;
    this.socChartKey = `socChartKey_${this.lastKey}`;
    this.timeout = setTimeout(() => this.reloadComponents(), 5000);
    console.log("****current", JSON.stringify(this.currentVehicleData));
  }

  mounted() {
    this.timeout = setTimeout(() => this.reloadComponents(), 5000);
  }

  beforeDestroy() {
    clearTimeout(this.timeout);
  }
}
</script>

<style lang="scss">
@import "@/assets/sass/variables";
$margin-left: 2rem;
$component-width: 70%;
$component-height: 14rem;

.main-container {
  display: flex;
  justify-content: flex-start;
  flex-direction: column;
  width: 97%;
  margin: 0 2rem;
  padding: 0;
  overflow: hidden;
}

.vehicle-container {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  height: 47%;
  width: 100%;
  margin: 0;
  padding: 0;
  background-color: transparent;

  @media screen and (min-width: $tablet-screen) {
    margin: 0 2rem;
  }

  @media screen and (min-width: $medium-screen) {
    flex-direction: row;
    width: 97%;
    margin: 0 1rem;
    padding: 1rem 0;
  }
}

.single-component {
  display: flex;
  height: $component-height;
  width: $component-width;
  margin: 1rem $margin-left 0 $margin-left;
  padding: 0;
  border-radius: 0.5rem;
  border: 2px solid $background-color;
  border-radius: 0.5rem;

  @media screen and (min-width: $medium-screen) {
    flex-direction: row;
    width: 50%;
    margin: 0 2rem 0 0;
  }
}
</style>
