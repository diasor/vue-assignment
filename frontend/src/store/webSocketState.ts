import Vue from "vue";
import Vuex from "vuex";
import { ActionTree, GetterTree, MutationTree } from "vuex";
import { RootState } from "@/types/rootState";
import { WebSocketState, Message } from "@/types/webSocketTypes";

Vue.use(Vuex);

const state: WebSocketState = {
  isConnected: false,
  messageList: [],
  connectionSocket: {}
};

// getters
export const getters: GetterTree<WebSocketState, RootState> = {
  messages: state => state.messageList,
  isConnected: state => state.isConnected
};

// mutations types
const mutationTypes = {
  CONNECT: "SOCKET_CONNECT",
  DISCONNECT: "SOCKET_DISCONNECT",
  ADD_ITEM_TO_CART_SUCCESS: "ADD_ITEM_TO_CART_SUCCESS",
  REMOVE_ITEM_FROM_CART: "REMOVE_ITEM_FROM_CART",
  REMOVE_ITEM_FROM_CART_SUCCESS: "REMOVE_ITEM_FROM_CART_SUCCESS"
};

const mutations: MutationTree<WebSocketState> = {
  CONNECT(state, socket) {
    console.log("mutation connect");
    state.connectionSocket = socket;
  },

  IS_CONNECTED(state, isConnected: boolean) {
    state.isConnected = isConnected;
  },

  DISCONNECT(state) {
    state.isConnected = false;
  },

  ADD_MESSAGE(state, message) {
    console.log("mutation socket_Data");
    // state.socketMessage = message;
    state.messageList.push(message);
  }
};

export const actions: ActionTree<WebSocketState, RootState> = {
  connect({ commit }) {
    // this method connects to the websocket where the backend sends the messasges
    const socket = new WebSocket("ws://localhost:3000");
    commit("CONNECT", socket);

    socket.onopen = function() {
      console.log("ws::open : connection established ");
      // this.status = 'Connected';
      commit("IS_CONNECTED", true);
    };

    socket.onerror = function(errorEvent: any) {
      console.log("WebSocket ERROR: " + JSON.stringify(errorEvent, null, 4));
      commit("IS_CONNECTED", false);
    };
  },

  getMessages({ commit, state }) {
    state.connectionSocket.onmessage = function(messageEvent: any) {
      const wsMessage = messageEvent.data;
      if (wsMessage.indexOf("error") > 0) {
        console.error("ws::msg_in:error: " + wsMessage.error);
      } else {
        console.info("ws::msg_in: " + wsMessage);
        commit("ADD_MESSAGE", wsMessage);
        console.log("cnt messages", state.messageList.length);
      }
    };
  }
};

export const webSocketState = {
  state,
  getters,
  actions,
  mutations,
  namespaced: true
};
