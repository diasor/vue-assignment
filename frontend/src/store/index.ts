import Vue from "vue";
import Vuex from "vuex";
import { webSocketState } from "@/store/webSocketState";

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    webSocketState
  }
});
