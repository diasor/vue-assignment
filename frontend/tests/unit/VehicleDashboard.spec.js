import { createLocalVue, mount, shallowMount } from "@vue/test-utils";
import Vuex from "vuex";
import VehicleDashboard from "@/components/vehicle-data/VehicleDashboard";
import { webSocketState } from "@/store/webSocketState";

const localVue = createLocalVue();
localVue.use(Vuex);

let getters;
let store;
let state;
let wrapper;

beforeEach(() => {
  state = {
    messageList: () => [
      {
        time: "1511437145000",
        energy: 56.283,
        gps: [52.08335494995117, 5.124546527862549],
        odo: 88529.27,
        speed: 21,
        soc: 64,
        parsedTime: "2017-11-23T11:39:05.000Z"
      },
      {
        time: "1511437398000",
        energy: 57.134,
        gps: [52.07914733886719, 5.119481563568115],
        odo: 88530.067,
        speed: 17,
        soc: 63.2,
        parsedTime: "2017-11-23T11:43:18.000Z"
      }
    ],
    currentVehicle: () => ({
      time: "1511437398000",
      energy: 57.134,
      gps: [52.07914733886719, 5.119481563568115],
      odo: 88530.067,
      speed: 17,
      soc: 63.2,
      parsedTime: "2017-11-23T11:43:18.000Z"
    })
  };

  getters = {
    messages: () => [
      {
        time: "1511437145000",
        energy: 56.283,
        gps: [52.08335494995117, 5.124546527862549],
        odo: 88529.27,
        speed: 21,
        soc: 64,
        parsedTime: "2017-11-23T11:39:05.000Z"
      },
      {
        time: "1511437398000",
        energy: 57.134,
        gps: [52.07914733886719, 5.119481563568115],
        odo: 88530.067,
        speed: 17,
        soc: 63.2,
        parsedTime: "2017-11-23T11:43:18.000Z"
      }
    ],

    currentVehicleData: () => ({
      time: "1511437398000",
      energy: 57.134,
      gps: [52.07914733886719, 5.119481563568115],
      odo: 88530.067,
      speed: 17,
      soc: 63.2,
      parsedTime: "2017-11-23T11:43:18.000Z"
    })
  };

  store = new Vuex.Store({
    modules: {
      webSocketState: {
        namespaced: true,
        state,
        getters: webSocketState.getters
      }
    }
  });

  // shallow mounting the main component with the store on a localVue instance and stubs for the subcomponents
  wrapper = shallowMount(VehicleDashboard, {
    store,
    localVue
  });
});

afterEach(() => {
  wrapper.destroy();
});

describe("VehicleDashboard.vue rendering", () => {
  // mounts the main dashboard component
  // on the process it tests the Vuex getters and the access to the module namespace
  it("renders component testing the Vuex getters and the access to the module namespace", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  // renders 1 main-container class, 2 vehicle-containers and 4 single-component
  it("has 1 main-container class, 2 vehicle-containers and 4 single-component", () => {
    expect(wrapper.findAll(".main-container").length).toBe(1);
    expect(wrapper.findAll(".vehicle-container").length).toBe(2);
    expect(wrapper.findAll(".single-component").length).toBe(4);
  });

  // tests that the main component renders 1 GoogleMap subcomponent
  it("renders 1 GoogleMap subcomponent", () => {
    expect(wrapper.find("google-map-stub").exists()).toBe(true);
  });

  // tests that the main component renders 1 VehicleData subcomponent
  it("renders 1 VehicleData subcomponent", () => {
    expect(wrapper.find("vehicle-data-stub").exists()).toBe(true);
  });

  // tests that the main component renders 2 LineCharts subcomponent
  it("renders 2 LineChart subcomponents", () => {
    expect(wrapper.findAll("line-chart-stub").length).toBe(2);
  });
});

describe("VehicleDashboard.vue Getters values", () => {
  // tests that the currentVehicleData got on the component is the proper one
  it("has a getter currentVehicleData equal to the one on the state", () => {
    expect(getters.currentVehicleData(state)).toEqual({
      time: "1511437398000",
      energy: 57.134,
      gps: [52.07914733886719, 5.119481563568115],
      odo: 88530.067,
      speed: 17,
      soc: 63.2,
      parsedTime: "2017-11-23T11:43:18.000Z"
    });
  });

  // tests that the vehicleData computed property has the value of the getter
  it("has a computed property vehicleData with the value of the state getter currentVehicleData", () => {
    expect(wrapper.vm.vehicleData()).toEqual(getters.currentVehicleData(state));
  });

  // tests that the messages got on the component has the proper length
  it("has a getter messages with a length of 2", () => {
    expect(getters.messages(state).length).toBe(2);
  });

  // tests that the allMessages computed property has the value of the getter "messages"
  it("has a computed property allMessages with the value of the state getter messages", () => {
    expect(wrapper.vm.allMessages()).toEqual(getters.messages(state));
  });
});
