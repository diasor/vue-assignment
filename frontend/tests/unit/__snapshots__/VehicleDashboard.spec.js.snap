// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleDashboard.vue rendering renders component testing the Vuex getters and the access to the module namespace 1`] = `
<div
  class="main-container"
>
  <div
    class="vehicle-container"
  >
    <div
      class="single-component"
    >
      <google-map-stub
        currentposition="function currentVehicle() {
      return {
        time: \\"1511437398000\\",
        energy: 57.134,
        gps: [52.07914733886719, 5.119481563568115],
        odo: 88530.067,
        speed: 17,
        soc: 63.2,
        parsedTime: \\"2017-11-23T11:43:18.000Z\\"
      };
    }"
        previouspositions="function messageList() {
      return [{
        time: \\"1511437145000\\",
        energy: 56.283,
        gps: [52.08335494995117, 5.124546527862549],
        odo: 88529.27,
        speed: 21,
        soc: 64,
        parsedTime: \\"2017-11-23T11:39:05.000Z\\"
      }, {
        time: \\"1511437398000\\",
        energy: 57.134,
        gps: [52.07914733886719, 5.119481563568115],
        odo: 88530.067,
        speed: 17,
        soc: 63.2,
        parsedTime: \\"2017-11-23T11:43:18.000Z\\"
      }];
    }"
      />
    </div>
     
    <div
      class="single-component"
    >
      <vehicle-data-stub
        energy="0"
        odo="0"
        soc="0"
        speed="0"
      />
    </div>
  </div>
   
  <div
    class="vehicle-container"
  >
    <div
      class="single-component"
    >
      <line-chart-stub
        data="function messageList() {
      return [{
        time: \\"1511437145000\\",
        energy: 56.283,
        gps: [52.08335494995117, 5.124546527862549],
        odo: 88529.27,
        speed: 21,
        soc: 64,
        parsedTime: \\"2017-11-23T11:39:05.000Z\\"
      }, {
        time: \\"1511437398000\\",
        energy: 57.134,
        gps: [52.07914733886719, 5.119481563568115],
        odo: 88530.067,
        speed: 17,
        soc: 63.2,
        parsedTime: \\"2017-11-23T11:43:18.000Z\\"
      }];
    }"
        datakey="speed"
        title="Speed Profile"
        xtitle="Time"
        ytitle="Speed (km/h)"
      />
    </div>
     
    <div
      class="single-component"
    >
      <line-chart-stub
        data="function messageList() {
      return [{
        time: \\"1511437145000\\",
        energy: 56.283,
        gps: [52.08335494995117, 5.124546527862549],
        odo: 88529.27,
        speed: 21,
        soc: 64,
        parsedTime: \\"2017-11-23T11:39:05.000Z\\"
      }, {
        time: \\"1511437398000\\",
        energy: 57.134,
        gps: [52.07914733886719, 5.119481563568115],
        odo: 88530.067,
        speed: 17,
        soc: 63.2,
        parsedTime: \\"2017-11-23T11:43:18.000Z\\"
      }];
    }"
        datakey="soc"
        title="State Of Charge Profile"
        xtitle="Time"
        ytitle="SoC (%)"
      />
    </div>
  </div>
</div>
`;
